<a href="https://colab.research.google.com/github/Don-Ho25/Intro_to_Predictive_Analysis/blob/main/Lesson1/Assignment1_DataEngineeringFundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import random
import os
from datetime import datetime

# Set random seed for reproducibility
random.seed(42)

In [4]:
print("CSV TO PARQUET CONVERSION AND ANALYSIS")
print("=" * 60)
print("  ")

CSV TO PARQUET CONVERSION AND ANALYSIS
  


In [5]:
# Step 1: Create automated dataset with 100+ records
print("\nStep 1: Creating automated dataset...")

# Define lists for random selection
names = [
    "Amara", "Kwame", "Zara", "Kofi", "Asha",
    "Jabari", "Kesi", "Tariq", "Nia", "Sekou"
]

cities = [
    "Beijing", "Shanghai", "Guangzhou", "Shenzhen", "Chengdu",
    "Hangzhou", "Wuhan", "Xi'an", "Suzhou", "Nanjing"
]

# Generate 100 records automatically
print("Generating 100 fictional user records...")
dataset = []

for i in range(100):
    record = {
        'Name': random.choice(names),
        'Age': random.randint(17, 65),
        'City': random.choice(cities)
    }
    dataset.append(record)

# Create DataFrame
df = pd.DataFrame(dataset)

print(f"Dataset created with {len(df)} records")
print("\nFirst 10 records:")
print(df.head(10))

print(f"\nDataset statistics:")
print(f"- Total records: {len(df)}")
print(f"- Unique names: {df['Name'].nunique()}")
print(f"- Unique cities: {df['City'].nunique()}")
print(f"- Age range: {df['Age'].min()} - {df['Age'].max()}")



Step 1: Creating automated dataset...
Generating 100 fictional user records...
Dataset created with 100 records

First 10 records:
    Name  Age      City
0  Amara   64  Hangzhou
1  Amara   20   Nanjing
2  Tariq   49    Suzhou
3   Zara   20    Suzhou
4  Kwame   28  Shanghai
5  Sekou   21  Shenzhen
6   Kesi   24   Nanjing
7   Kofi   54   Nanjing
8  Amara   56  Shanghai
9   Kesi   59   Nanjing

Dataset statistics:
- Total records: 100
- Unique names: 10
- Unique cities: 10
- Age range: 17 - 64


In [6]:

# Step 2: Save as CSV and convert to Parquet
print("\nStep 2: Converting CSV to Parquet...")

csv_filename = 'user_data.csv'
parquet_filename = 'user_data.parquet'

# Save as CSV
df.to_csv(csv_filename, index=False)
print(f"✓ CSV file saved: {csv_filename}")

# Convert to Parquet
df.to_parquet(parquet_filename, index=False)
print(f"✓ Parquet file saved: {parquet_filename}")



Step 2: Converting CSV to Parquet...
✓ CSV file saved: user_data.csv
✓ Parquet file saved: user_data.parquet


In [7]:
# Step 3: Compare file sizes
print("\nStep 3: File size comparison...")

csv_size = os.path.getsize(csv_filename)
parquet_size = os.path.getsize(parquet_filename)

print(f"CSV file size: {csv_size:,} bytes ({csv_size/1024:.2f} KB)")
print(f"Parquet file size: {parquet_size:,} bytes ({parquet_size/1024:.2f} KB)")

size_difference = csv_size - parquet_size
compression_ratio = (size_difference / csv_size) * 100

print(f"Size difference: {size_difference:,} bytes")
print(f"Compression ratio: {compression_ratio:.1f}% smaller")



Step 3: File size comparison...
CSV file size: 1,647 bytes (1.61 KB)
Parquet file size: 2,612 bytes (2.55 KB)
Size difference: -965 bytes
Compression ratio: -58.6% smaller


## **Analysis**

Contrary to typical expectations where Parquet files often offer significant compression advantages over CSV, in this specific instance, the Parquet file is actually larger than the CSV file by 965 bytes. This results in a negative "compression ratio" of −58.6%, indicating that the Parquet file is approximately 58.6% larger than the CSV file.

In [10]:
# Step 4: Load Parquet file and verify data integrity
print("\nStep 4: Loading Parquet file and verifying data integrity...")

df_parquet = pd.read_parquet(parquet_filename)

# Verify data integrity
print(f"Original DataFrame shape: {df.shape}")
print(f"Parquet DataFrame shape: {df_parquet.shape}")

# Check if data is identical
data_identical = df.equals(df_parquet)
print(f"Data integrity check: {' PASSED' if data_identical else ' FAILED'}")

if data_identical:
    print("All data successfully preserved during conversion!")
else:
    print("WARNING: Data differences detected!")

# Additional verification
print("\nDetailed verification:")
print(f"- Row count match: {len(df) == len(df_parquet)}")
print(f"- Column count match: {len(df.columns) == len(df_parquet.columns)}")
print(f"- Column names match: {list(df.columns) == list(df_parquet.columns)}")

# Sample comparison
print("\nSample data comparison (first 5 rows):")
print("ORIGINAL CSV DATA:")
print(df.head())
print("\nPARQUET DATA:")
print(df_parquet.head())



Step 4: Loading Parquet file and verifying data integrity...
Original DataFrame shape: (100, 3)
Parquet DataFrame shape: (100, 3)
Data integrity check:  PASSED
All data successfully preserved during conversion!

Detailed verification:
- Row count match: True
- Column count match: True
- Column names match: True

Sample data comparison (first 5 rows):
ORIGINAL CSV DATA:
    Name  Age      City
0  Amara   64  Hangzhou
1  Amara   20   Nanjing
2  Tariq   49    Suzhou
3   Zara   20    Suzhou
4  Kwame   28  Shanghai

PARQUET DATA:
    Name  Age      City
0  Amara   64  Hangzhou
1  Amara   20   Nanjing
2  Tariq   49    Suzhou
3   Zara   20    Suzhou
4  Kwame   28  Shanghai


In [12]:
# Step 5: Create comprehensive report
print("\n" + "=" * 60)
print("COMPREHENSIVE ANALYSIS REPORT")
print("=" * 60)

print(f"""
DATASET SUMMARY:
- Records generated: {len(df)}
- Data generation method: Automated using random.choice()
- Name pool: {len(names)} unique names
- City pool: {len(cities)} unique cities
- Age range: {df['Age'].min()} to {df['Age'].max()} years

FILE FORMAT COMPARISON:
- CSV file size: {csv_size:,} bytes ({csv_size/1024:.2f} KB)
- Parquet file size: {parquet_size:,} bytes ({parquet_size/1024:.2f} KB)
- Size reduction: {abs(size_difference):,} bytes ({abs(compression_ratio):.1f}%)
- Parquet is {'smaller' if parquet_size < csv_size else 'larger'} than CSV

DATA INTEGRITY:
- Data preservation: {'✓ Complete' if data_identical else '✗ Issues detected'}
- All records intact: {len(df) == len(df_parquet)}
- Schema preserved: {list(df.columns) == list(df_parquet.columns)}
""")

print("OBSERVATIONS:")
observations = []

if parquet_size < csv_size:
    observations.append(f"• Parquet format achieved {compression_ratio:.1f}% size reduction")
    observations.append("• Parquet's columnar storage is more efficient for this dataset")
else:
    observations.append("• CSV format is smaller for this particular dataset")
    observations.append("• Small datasets may not benefit from Parquet compression")

observations.append(f"• Dataset contains {df['Name'].nunique()} unique names across {len(df)} records")
observations.append(f"• Name distribution: {dict(df['Name'].value_counts().head(3))}")
observations.append(f"• City distribution: {dict(df['City'].value_counts().head(3))}")
observations.append(f"• Average age: {df['Age'].mean():.1f} years")

for obs in observations:
    print(obs)

print(f"\nRECOMMENDations:")
print("• Parquet format recommended for:")
print("  - Large datasets (>1000 records)")
print("  - Analytical workloads")
print("  - Long-term storage")
print("  - When file size matters")
print("• CSV format recommended for:")
print("  - Small datasets")
print("  - Human readability")
print("  - Simple data exchange")
print("  - When compatibility is crucial")


COMPREHENSIVE ANALYSIS REPORT

DATASET SUMMARY:
- Records generated: 100
- Data generation method: Automated using random.choice()
- Name pool: 10 unique names
- City pool: 10 unique cities
- Age range: 17 to 64 years

FILE FORMAT COMPARISON:
- CSV file size: 1,647 bytes (1.61 KB)
- Parquet file size: 2,612 bytes (2.55 KB)
- Size reduction: 965 bytes (58.6%)
- Parquet is larger than CSV

DATA INTEGRITY:
- Data preservation: ✓ Complete
- All records intact: True
- Schema preserved: True

OBSERVATIONS:
• CSV format is smaller for this particular dataset
• Small datasets may not benefit from Parquet compression
• Dataset contains 10 unique names across 100 records
• Name distribution: {'Asha': np.int64(15), 'Zara': np.int64(12), 'Kofi': np.int64(12)}
• City distribution: {'Suzhou': np.int64(14), 'Hangzhou': np.int64(12), 'Shanghai': np.int64(12)}
• Average age: 40.3 years

RECOMMENDations:
• Parquet format recommended for:
  - Large datasets (>1000 records)
  - Analytical workloads
  - L

In [14]:
import pandas as pd
import random
from datetime import datetime

# Set random seed for reproducibility
random.seed(42)

print("=" * 60)
print("BASIC ETL PIPELINE IMPLEMENTATION")
print("=" * 60)

# First, let's create sample data for demonstration
print("Creating sample dataset for ETL demonstration...")

# Create diverse sample data with some under-18 records for filtering
sample_data = {
    'Name': [
        'Amara Johnson', 'Kwame Smith', 'Zara Brown', 'Kofi Davis', 'Asha Wilson',
        'Jabari Miller', 'Kesi Jones', 'Tariq Garcia', 'Nia Rodriguez', 'Sekou Martinez',
        'Fatima Lopez', 'Chinonso Hernandez', 'Adunni Anderson', 'Olumide Taylor',
        'Chioma Thomas', 'Emeka Jackson', 'Ngozi White', 'Adebayo Harris',
        'Funmi Martin', 'Tunde Thompson'
    ],
    'Age': [
        25, 17, 30, 16, 28, 35, 15, 22, 31, 19,
        24, 18, 27, 14, 33, 29, 20, 26, 32, 21
    ],
    'City': [
        'Beijing', 'Shanghai', 'Guangzhou', 'Shenzhen', 'Chengdu',
        'Hangzhou', 'Wuhan', 'Xi\'an', 'Suzhou', 'Nanjing',
        'Beijing', 'Shanghai', 'Guangzhou', 'Shenzhen', 'Chengdu',
        'Hangzhou', 'Wuhan', 'Xi\'an', 'Suzhou', 'Nanjing'
    ],
    'Salary': [
        50000, 35000, 75000, 28000, 65000, 85000, 25000, 45000, 72000, 42000,
        55000, 38000, 68000, 22000, 90000, 78000, 48000, 62000, 80000, 52000
    ]
}

# Create DataFrame and save as source CSV
source_df = pd.DataFrame(sample_data)
source_file = 'source_data.csv'
source_df.to_csv(source_file, index=False)

print(f"Sample dataset created with {len(source_df)} records")
print(f"Source file saved: {source_file}")

# ETL PIPELINE IMPLEMENTATION
print("\n" + "=" * 60)
print("STARTING ETL PIPELINE")
print("=" * 60)

# STEP 1: EXTRACT DATA FROM CSV FILE
print("\nSTEP 1: EXTRACT - Loading data from CSV file")
print("-" * 40)

try:
    # Load data from CSV file
    extracted_df = pd.read_csv(source_file)
    print(f"✓ Successfully extracted {len(extracted_df)} records from {source_file}")
    print(f"✓ Columns: {list(extracted_df.columns)}")

    print("\nExtracted data preview:")
    print(extracted_df.head(10))

    # Data quality check
    print(f"\nData quality summary:")
    print(f"- Total records: {len(extracted_df)}")
    print(f"- Records with age < 18: {len(extracted_df[extracted_df['Age'] < 18])}")
    print(f"- Records with age >= 18: {len(extracted_df[extracted_df['Age'] >= 18])}")
    print(f"- Missing values: {extracted_df.isnull().sum().sum()}")

except Exception as e:
    print(f"✗ Error during extraction: {e}")
    exit(1)

# STEP 2: TRANSFORM DATA
print("\nSTEP 2: TRANSFORM - Filtering and converting data")
print("-" * 40)

try:
    # Create a copy for transformation
    transformed_df = extracted_df.copy()

    # Show original data statistics
    print(f"Before transformation:")
    print(f"- Total records: {len(transformed_df)}")
    print(f"- Records with age < 18: {len(transformed_df[transformed_df['Age'] < 18])}")
    print(f"- Sample names: {transformed_df['Name'].head(3).tolist()}")

    # Transformation 1: Filter out records where age is below 18
    print("\nApplying transformation 1: Filtering age >= 18...")
    records_before_filter = len(transformed_df)
    transformed_df = transformed_df[transformed_df['Age'] >= 18]
    records_after_filter = len(transformed_df)
    records_filtered = records_before_filter - records_after_filter

    print(f"✓ Filtered out {records_filtered} records with age < 18")
    print(f"✓ Remaining records: {records_after_filter}")

    # Transformation 2: Convert names to uppercase
    print("\nApplying transformation 2: Converting names to uppercase...")
    transformed_df['Name'] = transformed_df['Name'].str.upper()
    print(f"✓ All names converted to uppercase")
    print(f"✓ Sample transformed names: {transformed_df['Name'].head(3).tolist()}")

    print(f"\nAfter transformation:")
    print(f"- Total records: {len(transformed_df)}")
    print(f"- All ages >= 18: {all(transformed_df['Age'] >= 18)}")
    print(f"- All names uppercase: {all(name.isupper() for name in transformed_df['Name'])}")

except Exception as e:
    print(f"✗ Error during transformation: {e}")
    exit(1)

# STEP 3: LOAD TRANSFORMED DATA INTO NEW CSV FILE
print("\nSTEP 3: LOAD - Saving transformed data to new CSV file")
print("-" * 40)

try:
    # Save transformed data to new CSV file
    output_file = 'transformed_data.csv'
    transformed_df.to_csv(output_file, index=False)

    print(f"✓ Transformed data saved to {output_file}")
    print(f"✓ File contains {len(transformed_df)} records")

    # Check file was created successfully
    import os
    if os.path.exists(output_file):
        file_size = os.path.getsize(output_file)
        print(f"✓ File size: {file_size} bytes")

except Exception as e:
    print(f"✗ Error during loading: {e}")
    exit(1)

# STEP 4: VERIFY TRANSFORMATION
print("\nSTEP 4: VERIFY - Loading and displaying the new file")
print("-" * 40)

try:
    # Load the transformed file to verify
    verification_df = pd.read_csv(output_file)

    print(f"✓ Successfully loaded {len(verification_df)} records from {output_file}")

    # Verification checks
    print("\nVerification Results:")
    print(f"- Records in transformed file: {len(verification_df)}")
    print(f"- All ages >= 18: {all(verification_df['Age'] >= 18)}")
    print(f"- All names uppercase: {all(name.isupper() for name in verification_df['Name'])}")
    print(f"- Columns preserved: {list(verification_df.columns) == list(extracted_df.columns)}")

    # Display transformed data
    print("\nTransformed data (first 10 records):")
    print(verification_df.head(10))

    # Summary statistics
    print(f"\nSummary statistics:")
    print(f"- Age range: {verification_df['Age'].min()} - {verification_df['Age'].max()}")
    print(f"- Average age: {verification_df['Age'].mean():.1f}")
    print(f"- Unique cities: {verification_df['City'].nunique()}")

except Exception as e:
    print(f"✗ Error during verification: {e}")
    exit(1)

# STEP 5: FINAL SUMMARY AND FILE CONFIRMATION
print("\nSTEP 5: FINAL SUMMARY - ETL Pipeline Complete")
print("-" * 40)

print(f"""
ETL PIPELINE EXECUTION SUMMARY:
{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}

✓ EXTRACT: Loaded {len(extracted_df)} records from {source_file}
✓ TRANSFORM:
  - Filtered {records_filtered} records with age < 18
  - Converted all names to uppercase
  - Final dataset: {len(transformed_df)} records
✓ LOAD: Saved transformed data to {output_file}
✓ VERIFY: Confirmed data integrity and transformations

FILES CREATED:
- Source file: {source_file} ({len(source_df)} records)
- Transformed file: {output_file} ({len(verification_df)} records)

TRANSFORMATION SUMMARY:
- Records removed (age < 18): {records_filtered}
- Records retained: {len(verification_df)}
- Names converted to uppercase: {len(verification_df)}
- Data integrity: ✓ VERIFIED

The ETL pipeline has been successfully completed!
""")

# Display side-by-side comparison
print("\nBEFORE vs AFTER COMPARISON:")
print("=" * 60)
print("ORIGINAL DATA (first 5 records):")
print(extracted_df.head())
print("\nTRANSFORMED DATA (first 5 records):")
print(verification_df.head())

BASIC ETL PIPELINE IMPLEMENTATION
Creating sample dataset for ETL demonstration...
Sample dataset created with 20 records
Source file saved: source_data.csv

STARTING ETL PIPELINE

STEP 1: EXTRACT - Loading data from CSV file
----------------------------------------
✓ Successfully extracted 20 records from source_data.csv
✓ Columns: ['Name', 'Age', 'City', 'Salary']

Extracted data preview:
             Name  Age       City  Salary
0   Amara Johnson   25    Beijing   50000
1     Kwame Smith   17   Shanghai   35000
2      Zara Brown   30  Guangzhou   75000
3      Kofi Davis   16   Shenzhen   28000
4     Asha Wilson   28    Chengdu   65000
5   Jabari Miller   35   Hangzhou   85000
6      Kesi Jones   15      Wuhan   25000
7    Tariq Garcia   22      Xi'an   45000
8   Nia Rodriguez   31     Suzhou   72000
9  Sekou Martinez   19    Nanjing   42000

Data quality summary:
- Total records: 20
- Records with age < 18: 4
- Records with age >= 18: 16
- Missing values: 0

STEP 2: TRANSFORM - Filt

In [19]:
import sqlite3
import json
from datetime import datetime
import os


In [21]:
print("SQLITE vs MONGODB COMPARISON")
print("=" * 70)


SQLITE vs MONGODB COMPARISON


In [22]:
# Sample data for both databases
sample_users = [
    {"ID": 1, "Name": "Amara Johnson", "Age": 25, "City": "Beijing", "Department": "Engineering", "Salary": 75000},
    {"ID": 2, "Name": "Kwame Smith", "Age": 30, "City": "Shanghai", "Department": "Marketing", "Salary": 65000},
    {"ID": 3, "Name": "Zara Brown", "Age": 28, "City": "Guangzhou", "Department": "Finance", "Salary": 70000},
    {"ID": 4, "Name": "Kofi Davis", "Age": 32, "City": "Shenzhen", "Department": "HR", "Salary": 60000},
    {"ID": 5, "Name": "Asha Wilson", "Age": 27, "City": "Chengdu", "Department": "Engineering", "Salary": 78000},
    {"ID": 6, "Name": "Jabari Miller", "Age": 35, "City": "Hangzhou", "Department": "Sales", "Salary": 72000}
]

# Additional flexible data for MongoDB (showing document model flexibility)
flexible_users = [
    {
        "ID": 7, "Name": "Fatima Chen", "Age": 29, "City": "Wuhan",
        "Department": "Engineering", "Salary": 80000,
        "Skills": ["Python", "JavaScript", "React"],
        "Projects": [
            {"name": "E-commerce Platform", "status": "completed"},
            {"name": "Mobile App", "status": "in_progress"}
        ]
    },
    {
        "ID": 8, "Name": "Chinonso Wang", "Age": 31, "City": "Xi'an",
        "Department": "Marketing", "Salary": 68000,
        "Skills": ["SEO", "Content Marketing"],
        "Certifications": ["Google Analytics", "Facebook Ads"]
    }
]



In [24]:
print("PART 1: SQLite RELATIONAL DATABASE OPERATIONS")
print("=" * 70)

PART 1: SQLite RELATIONAL DATABASE OPERATIONS


In [25]:
def sqlite_operations():
    """Demonstrate SQLite relational database operations"""

    # Create/connect to SQLite database
    db_file = 'users.db'
    print(f"\n1. Creating SQLite database: {db_file}")

    try:
        conn = sqlite3.connect(db_file)
        cursor = conn.cursor()

        # Create table
        print("2. Creating users table...")
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS users (
                ID INTEGER PRIMARY KEY,
                Name TEXT NOT NULL,
                Age INTEGER NOT NULL,
                City TEXT NOT NULL,
                Department TEXT,
                Salary REAL
            )
        ''')

        # Clear existing data for fresh start
        cursor.execute("DELETE FROM users")

        # Insert sample data
        print("3. Inserting sample records...")
        for user in sample_users:
            cursor.execute('''
                INSERT INTO users (ID, Name, Age, City, Department, Salary)
                VALUES (?, ?, ?, ?, ?, ?)
            ''', (user['ID'], user['Name'], user['Age'], user['City'],
                  user['Department'], user['Salary']))

        conn.commit()
        print(f"✓ Inserted {len(sample_users)} records into SQLite")

        # Retrieve and display data
        print("\n4. Retrieving data using SQL queries:")

        # Query 1: All users
        print("\nQuery 1: All users")
        cursor.execute("SELECT * FROM users")
        results = cursor.fetchall()

        print(f"{'ID':<3} {'Name':<15} {'Age':<4} {'City':<12} {'Department':<12} {'Salary':<8}")
        print("-" * 65)
        for row in results:
            print(f"{row[0]:<3} {row[1]:<15} {row[2]:<4} {row[3]:<12} {row[4]:<12} {row[5]:<8}")

        # Query 2: Users older than 29
        print("\nQuery 2: Users older than 29")
        cursor.execute("SELECT * FROM users WHERE Age > 29")
        results = cursor.fetchall()

        print("\nUsers older than 29:")
        print(f"{'ID':<3} {'Name':<15} {'Age':<4} {'City':<12} {'Department':<12} {'Salary':<8}")
        print("-" * 65)
        for row in results:
            print(f"{row[0]:<3} {row[1]:<15} {row[2]:<4} {row[3]:<12} {row[4]:<12} {row[5]:<8}")

        # Query 3: Average salary by department
        print("\nQuery 3: Average salary by department")
        cursor.execute("SELECT Department, AVG(Salary) FROM users GROUP BY Department")
        results = cursor.fetchall()

        print("\nAverage salary by department:")
        print(f"{'Department':<15} {'Average Salary':<15}")
        print("-" * 30)
        for row in results:
            print(f"{row[0]:<15} {row[1]:<15.2f}")

    except sqlite3.Error as e:
        print(f"✗ SQLite error: {e}")
    finally:
        if 'conn' in locals() and conn:
            conn.close()
            print(f"\nSQLite connection closed.")


In [26]:
print("PART 2: MongoDB DOCUMENT DATABASE OPERATIONS")
print("=" * 70)


PART 2: MongoDB DOCUMENT DATABASE OPERATIONS


In [27]:

def mongodb_operations():
    """Demonstrate MongoDB document database operations"""

    if not MONGODB_AVAILABLE:
        print("MongoDB operations skipped: PyMongo not installed.")
        return

    # Connect to MongoDB
    try:
        print("\n1. Connecting to MongoDB...")
        # Replace 'mongodb://localhost:27017/' with your MongoDB connection string if needed
        client = MongoClient('mongodb://localhost:27017/', serverSelectionTimeoutMS=5000)
        client.admin.command('ping') # Check connection
        print("✓ Connected to MongoDB!")

        db = client['user_database'] # Use a database named 'user_database'
        collection = db['users'] # Use a collection named 'users'

        # Clear existing data for fresh start
        print("2. Clearing existing records in MongoDB collection...")
        delete_result = collection.delete_many({})
        print(f"✓ Deleted {delete_result.deleted_count} existing records")

        # Insert sample data (including flexible data)
        print("3. Inserting sample records...")
        all_users_data = sample_users + flexible_users
        insert_result = collection.insert_many(all_users_data)
        print(f"✓ Inserted {len(insert_result.inserted_ids)} records into MongoDB")

        # Retrieve and display data
        print("\n4. Retrieving data using MongoDB queries:")

        # Query 1: All users
        print("\nQuery 1: All users")
        results = list(collection.find({}).limit(10)) # Limit to 10 for preview

        print("\nAll users (first 10):")
        for user in results:
            # Print relevant fields, handling flexible schema
            print(f"ID: {user.get('ID')}, Name: {user.get('Name')}, Age: {user.get('Age')}, City: {user.get('City')}, Department: {user.get('Department')}, Salary: {user.get('Salary')}")
            if 'Skills' in user:
                print(f"  Skills: {user['Skills']}")
            if 'Projects' in user:
                print(f"  Projects: {user['Projects']}")
            if 'Certifications' in user:
                print(f"  Certifications: {user['Certifications']}")
            print("-" * 20)


        # Query 2: Users older than 29
        print("\nQuery 2: Users older than 29")
        results = list(collection.find({"Age": {"$gt": 29}}))

        print("\nUsers older than 29:")
        for user in results:
             print(f"ID: {user.get('ID')}, Name: {user.get('Name')}, Age: {user.get('Age')}, City: {user.get('City')}")

        # Query 3: Users with 'Engineering' department
        print("\nQuery 3: Users in Engineering department")
        results = list(collection.find({"Department": "Engineering"}))

        print("\nUsers in Engineering department:")
        for user in results:
             print(f"ID: {user.get('ID')}, Name: {user.get('Name')}, Age: {user.get('Age')}, City: {user.get('City')}")

        # Query 4: Users with 'Python' skill (demonstrating querying embedded arrays)
        print("\nQuery 4: Users with 'Python' skill")
        results = list(collection.find({"Skills": "Python"}))

        print("\nUsers with 'Python' skill:")
        for user in results:
             print(f"ID: {user.get('ID')}, Name: {user.get('Name')}, Age: {user.get('Age')}, City: {user.get('City')}, Skills: {user.get('Skills')}")


    except ConnectionFailure as e:
        print(f"✗ MongoDB connection error: Could not connect to MongoDB server. Please ensure MongoDB is running and accessible.")
        print(f"Error details: {e}")
    except Exception as e:
        print(f"✗ MongoDB error: {e}")
    finally:
        if 'client' in locals() and client:
            client.close()
            print(f"\nMongoDB connection closed.")


In [28]:
# Call sqlite_operations
sqlite_operations()


1. Creating SQLite database: users.db
2. Creating users table...
3. Inserting sample records...
✓ Inserted 6 records into SQLite

4. Retrieving data using SQL queries:

Query 1: All users
ID  Name            Age  City         Department   Salary  
-----------------------------------------------------------------
1   Amara Johnson   25   Beijing      Engineering  75000.0 
2   Kwame Smith     30   Shanghai     Marketing    65000.0 
3   Zara Brown      28   Guangzhou    Finance      70000.0 
4   Kofi Davis      32   Shenzhen     HR           60000.0 
5   Asha Wilson     27   Chengdu      Engineering  78000.0 
6   Jabari Miller   35   Hangzhou     Sales        72000.0 

Query 2: Users older than 29

Users older than 29:
ID  Name            Age  City         Department   Salary  
-----------------------------------------------------------------
2   Kwame Smith     30   Shanghai     Marketing    65000.0 
4   Kofi Davis      32   Shenzhen     HR           60000.0 
6   Jabari Miller   35   Ha

In [29]:
# Call MONGODB_operations
mongodb_operations()


1. Connecting to MongoDB...
✗ MongoDB connection error: Could not connect to MongoDB server. Please ensure MongoDB is running and accessible.
Error details: localhost:27017: [Errno 111] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms), Timeout: 5.0s, Topology Description: <TopologyDescription id: 6872ce9de6f147db4b042fbd, topology_type: Unknown, servers: [<ServerDescription ('localhost', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('localhost:27017: [Errno 111] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>

MongoDB connection closed.


COULD NOT work with mongobd

In [30]:
!jupyter nbconvert --to html  /content/Assignment1_DataEngineeringFundamentals.ipynb

[NbConvertApp] Converting notebook /content/Assignment1_DataEngineeringFundamentals.ipynb to html
[NbConvertApp] Writing 399439 bytes to /content/Assignment1_DataEngineeringFundamentals.html
